process/block/preprocess/transactions.go

```diff
@@ -220,65 +232,249 @@ func (txs *transactions) ProcessBlockTransactions(
+			TxHash:          txHash,
+			SenderShardID:   calculatedSenderShardId,
+			ReceiverShardID: calculatedReceiverShardId,
+		}
+
+		txsFromMiniBlock = append(txsFromMiniBlock, wrappedTx)
+	}
+
+	return txsFromMiniBlock, nil
+}
+
+func (txs *transactions) getShardFromAddress(address []byte) (uint32, error) {
+	isEmptyAddress := bytes.Equal(address, make([]byte, txs.addressConverter.AddressLen()))
+	if isEmptyAddress {
+		return txs.shardCoordinator.SelfId(), nil
+	}
+
+	addressContainer, err := txs.addressConverter.CreateAddressFromPublicKeyBytes(address)
+	if err != nil {
+		return 0, err
+	}
+
+	return txs.shardCoordinator.ComputeId(addressContainer), nil
+}
+
+func (txs *transactions) processTxsToMe(
+	body block.Body,
+	haveTime func() bool,
+) error {
+
+	startTime := time.Now()
+	defer func() {
+		elapsedTime := time.Since(startTime)
+		log.Debug("elapsed time to processTxsToMe",
+			"time [s]", elapsedTime,
+		)
+	}()
+
+	txsToMe, err := txs.computeTxsToMe(body)
+	if err != nil {
+		return err
+	}
+
+	//TODO: Should be verified if gas computation should be done here (before the next for) and should be done
+	//separately per miniblock, exactly as in the initial implementation. Also should be checked if with the last
+	//change, which call ProcessBlockTransactions method two times, first for mini blocks with destination in self shard
+	// and afterward with miniblocks from self shard, the gas computation is still correct.
+	gasConsumedByMiniBlockInSenderShard := uint64(0)
+	gasConsumedByMiniBlockInReceiverShard := uint64(0)
+	totalGasConsumedInSelfShard := txs.gasHandler.TotalGasConsumed()
+
+	for index := range txsToMe {
+		if !haveTime() {
+			return process.ErrTimeIsOut
+		}
+
+		txHandler := txsToMe[index].Tx
+		tx := txHandler.(*transaction.Transaction)
+		txHash := txsToMe[index].TxHash
 
-			txHash := miniBlock.TxHashes[j]
-			txs.txsForCurrBlock.mutTxsForBlock.RLock()
-			txInfoFromMap := txs.txsForCurrBlock.txHashAndInfo[string(txHash)]
+		txs.txsForCurrBlock.mutTxsForBlock.RLock()
+		txInfoFromMap := txs.txsForCurrBlock.txHashAndInfo[string(txHash)]
+		if txInfoFromMap == nil || txInfoFromMap.tx == nil {
 			txs.txsForCurrBlock.mutTxsForBlock.RUnlock()
+			log.Debug("missing transaction in processTxsToMe", "txHash", txHash)
+			return process.ErrMissingTransaction
+		}
 
-			if txInfoFromMap == nil || txInfoFromMap.tx == nil {
-				log.Debug("missing transaction in ProcessBlockTransactions ", "type", block.TxBlock, "txHash", txHash)
-				return process.ErrMissingTransaction
-			}
+		senderShardID := txInfoFromMap.senderShardID
+		receiverShardID := txInfoFromMap.receiverShardID
+		txs.txsForCurrBlock.mutTxsForBlock.RUnlock()
 
-			tx, ok := txInfoFromMap.tx.(*transaction.Transaction)
-			if !ok {
-				return process.ErrWrongTypeAssertion
-			}
+		err = txs.processAndRemoveBadTransaction(
+			txHash,
+			tx,
+			senderShardID,
+			receiverShardID,
+		)
+		if err != nil && !errors.Is(err, process.ErrFailedTransaction) {
```

Comment:
there is no transaction TO ME which is of type FAILED. so return err in case of any error

Reply:
Done